kakao/src/main/kotlin/com/agoda/kakao/screen/Screen.kt

```diff
@@ -24,15 +27,104 @@ import com.agoda.kakao.common.views.KBaseView
 @Suppress("UNCHECKED_CAST")
 @KakaoDslMarker
 open class Screen<out T: Screen<T>>: ScreenActions {
-    override val view: ViewInteraction = Espresso.onView(ViewMatchers.isRoot())
-    operator fun invoke(function: T.() -> Unit) = function.invoke(this as T)
+    private var vi: Interceptor<ViewInteraction, ViewAssertion, ViewAction>? = null
+    private var di: Interceptor<DataInteraction, ViewAssertion, ViewAction>? = null
+    private var wi: Interceptor<Web.WebInteraction<*>, WebAssertion<*>, Atom<*>>? = null
+
+    override val view: ViewInteractionDelegate = ViewInteractionDelegate(Espresso.onView(ViewMatchers.isRoot()))
+
+    private var isPushed = false
 
     /**
      * The visibility of rootView will be checked when entering the screen
-     * @rootView.isVisible() Will be called after land onScreen<>()
+     * @rootView.isVisible() Will be called after land onScreen<>() and after invoking screen directly.
      */
     open var rootView: KBaseView<*>? = null
 
+    /**
+     * Sets the interceptors for the screen.
+     * Interceptors will be invoked on all interactions while the screen is active.
+     *
+     * The screen is considered `active` when it is invoked in one of the following ways:
+     * ```
+     * val screen = SomeScreen()
+     *
+     * screen { // Active
+     *     view { click() }
+     *     ...
+     * } // Inactive
+     *
+     * // OR
+     *
+     * onScreen<SomeScreen>() { // Active
+     *     view { click() }
+     *     ...
+     * } // Inactive
+     * ```
+     *
+     * If you use nesting, interceptors of the latest screen that became active will be invoked.
+     * All active screen's interceptors are being stored in a stack, so that you won't.
+     *
+     * @param configurator Configuration of the interceptors
+     *
+     * @see Interceptor
+     */
+    fun intercept(configurator: Interceptor.Configurator.() -> Unit) {
+        var push = false
+
+        if (isPushed) {
+            push = true
+            pop()
```

Comment:
First. We use two boolean flags here. What problem do we try to resolve?
Second. If we call ```intercept``` second time in the Screen we remove the first intercept of the Screen. It's so non-obviously. 
If we want to prevent ```intercept``` method calling multiple times at concrete Screen then there is a thought to throw an exception?

Reply:
If `intercept` is called inside of the `onScreen<> {}` lambda and it had another interceptors being set in his `init` block. That means that we have to pop it's old interceptors and replace it with new ones. But `pop()` will reset the `isPushed` flag. I will anyway refactor this, so there won't be any flags.